### PR TITLE
Add mkdir_p to FileUtils.install

### LINKED ISSUE #103: FileUtils.install needs integrated mkdir_p
In the [Chromebrew project](https://github.com/chromebrew/chromebrew), we're making heavy use of `FileUtils.install` as a replacement for `system 'install...' `, but whereas the standalone `install` binary will automatically create necessary directories for an install, `FileUtils.install` will not.

Ideally `FileUtils.install` would internally also run `FileUtils.mkdir_p` before it installs a file into a destination, so we don't have to manually do such a step before running the `FileUtils.install` step.

### PR DESCRIPTION
**Pull request by satmandu**
Fixes #103 

- PR to solve the aforementioned issue of FileUtils.install not creating a destination directory before installing a file to a destination directory.
- Using `mkdir_p` on the destination directory solves that issue.

### CODE REVIEW

File: lib/fileutils.rb [hunk 1]
```diff
@@ -1642,6 +1642,12 @@ def install(src, dest, mode: nil, owner: nil, group: nil, preserve: nil,
       st = File.stat(s)
       unless File.exist?(d) and compare_file(s, d)
         remove_file d, true
+        begin
+          require 'pathname'
+        rescue LoadError
+          raise "cannot load pathname"
```
Comment (uberhacker) [inline lib/fileutils.rb:1648]: Would a more descriptive error message like `raise  "Unable to create #{Pathname.new(d).parent}.  Cannot load module pathname."` be more helpful?

Comment (satmandu) [reply lib/fileutils.rb:1648]: That doesn't work, as the raise message requires pathname to be loaded...

File: lib/fileutils.rb [hunk 2]
```diff
@@ -1642,6 +1642,12 @@ def install(src, dest, mode: nil, owner: nil, group: nil, preserve: nil,
       st = File.stat(s)
       unless File.exist?(d) and compare_file(s, d)
         remove_file d, true
+        begin
+          require 'pathname'
+        rescue LoadError
+          raise "Cannot load module pathname."
```
Comment (uberhacker) [inline lib/fileutils.rb:1648]: Yes, that was the wrong suggestion.  Can the input parameter for the directory be used instead?  I'm just thinking the current message is not as helpful for debugging.

Comment (satmandu) [reply lib/fileutils.rb]: How does the new verbiage sound?

Comment (uberhacker) [reply lib/fileutils.rb]: That is a bit more clear to hep debug.

File: lib/fileutils.rb [hunk 3]
```diff
@@ -1642,6 +1642,12 @@ def install(src, dest, mode: nil, owner: nil, group: nil, preserve: nil,
       st = File.stat(s)
       unless File.exist?(d) and compare_file(s, d)
         remove_file d, true
+        begin
+          require 'pathname'
+        rescue LoadError
+          raise "Cannot load module pathname for determining parent directory for #{d}."
+        end
+        mkdir_p Pathname.new(d).parent
```
[the remark this answers is not in the document]
Comment (satmandu) [reply lib/fileutils.rb:1650]: Just needs to handle it d ends with a trailing slash...

i.e. something with `if d.end_with?('/')`

[the remark this answers is not in the document]
Comment (supechicken) [reply lib/fileutils.rb:1650]: It also works with trailing slash
```ruby
File.expand_path('..', '/test//////') # => "/test"
```

[the remark this answers is not in the document]
Comment (satmandu) [reply lib/fileutils.rb:1650]: I was getting an error with `mkdir_p File.expand_path('..', d)`:
```
$ touch test
$ irb
irb(main):001:0> require './fileutils/lib/fileutils.rb'
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:183: warning: previous definition of VERSION was here
/localnet/src/fileutils/lib/fileutils.rb:2172: warning: already initialized constant FileUtils::Entry_::S_IF_DOOR
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2165: warning: previous definition of S_IF_DOOR was here
/localnet/src/fileutils/lib/fileutils.rb:2468: warning: already initialized constant FileUtils::Entry_::DIRECTORY_TERM
/localnet/src/fileutils/lib/fileutils.rb:2576: warning: already initialized constant FileUtils::OPT_TABLE
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2569: warning: previous definition of OPT_TABLE was here
/localnet/src/fileutils/lib/fileutils.rb:2634: warning: already initialized constant FileUtils::LOW_METHODS
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2627: warning: previous definition of LOW_METHODS was here
/localnet/src/fileutils/lib/fileutils.rb:2641: warning: already initialized constant FileUtils::METHODS
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2634: warning: previous definition of METHODS was here
=> true
irb(main):004:0> FileUtils.install 'test', 'tmp3/'
/localnet/src/fileutils/lib/fileutils.rb:2292:in `initialize': Is a directory @ rb_sysopen - tmp3/ (Errno::EISDIR)
        from /localnet/src/fileutils/lib/fileutils.rb:2292:in `open'
        from /localnet/src/fileutils/lib/fileutils.rb:2292:in `block in copy_file'
        from /localnet/src/fileutils/lib/fileutils.rb:2291:in `open'
        from /localnet/src/fileutils/lib/fileutils.rb:2291:in `copy_file'
        from /localnet/src/fileutils/lib/fileutils.rb:1088:in `copy_file'
        from /localnet/src/fileutils/lib/fileutils.rb:1646:in `block in install'
        from /localnet/src/fileutils/lib/fileutils.rb:2482:in `block in fu_each_src_dest'
        from /localnet/src/fileutils/lib/fileutils.rb:2480:in `fu_each_src_dest'
        from /localnet/src/fileutils/lib/fileutils.rb:1641:in `install'
        from (irb):4:in `<main>'
        from /var/lib/gems/3.0.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/bin/irb:25:in `load'
        from /usr/local/bin/irb:25:in `<main>'
irb(main):005:0>
```



This seems to work:
```
        if d.to_s.end_with?('/')
          puts "d is dir"
          mkdir_p d
          copy_file s, d + File.basename(s)
        else
          mkdir_p File.expand_path('..', d)
          copy_file s, d
        end
```

[the remark this answers is not in the document]
Comment (satmandu) [reply lib/fileutils.rb:1650]:
```
$ irb
irb(main):001:0> require './fileutils/lib/fileutils.rb'
/localnet/src/fileutils/lib/fileutils.rb:183: warning: already initialized constant FileUtils::VERSION
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:183: warning: previous definition of VERSION was here
/localnet/src/fileutils/lib/fileutils.rb:2171: warning: already initialized constant FileUtils::Entry_::S_IF_DOOR
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2165: warning: previous definition of S_IF_DOOR was here
/localnet/src/fileutils/lib/fileutils.rb:2467: warning: already initialized constant FileUtils::Entry_::DIRECTORY_TERM
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2461: warning: previous definition of DIRECTORY_TERM was here
/localnet/src/fileutils/lib/fileutils.rb:2575: warning: already initialized constant FileUtils::OPT_TABLE
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2569: warning: previous definition of OPT_TABLE was here
/localnet/src/fileutils/lib/fileutils.rb:2633: warning: already initialized constant FileUtils::LOW_METHODS
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2627: warning: previous definition of LOW_METHODS was here
/localnet/src/fileutils/lib/fileutils.rb:2640: warning: already initialized constant FileUtils::METHODS
/var/lib/gems/3.0.0/gems/fileutils-1.7.0/lib/fileutils.rb:2634: warning: previous definition of METHODS was here
=> true
irb(main):002:0>  FileUtils.install 'test', 'tmp20'
=> nil
irb(main):003:0>  FileUtils.install 'test', 'tmp21/'
=> nil
irb(main):004:0>  FileUtils.install 'test', 'tmp22/tmp'
=> nil
irb(main):005:0>  FileUtils.install 'test', 'tmp23/tmp/'
=> nil
irb(main):006:0> system "ls -aFl tmp21 ; ls -aFl tmp22 ; ls -aFl tmp23/tmp ; ls -aFl | grep tmp20"
total 51
drwxrwxr-x+   2 satadru satadru   3 Jan 11 11:59 ./
drwxrwxr-x+ 127 satadru satadru 231 Jan 11 11:59 ../
-rw-rw-r--+   1 satadru satadru   5 Jan 11 11:59 test
total 51
drwxrwxr-x+   2 satadru satadru   3 Jan 11 11:59 ./
drwxrwxr-x+ 127 satadru satadru 231 Jan 11 11:59 ../
-rw-rw-r--+   1 satadru satadru   5 Jan 11 11:59 tmp
total 3
drwxrwxr-x+ 2 satadru satadru 3 Jan 11 11:59 ./
drwxrwxr-x+ 3 satadru satadru 3 Jan 11 11:59 ../
-rw-rw-r--+ 1 satadru satadru 5 Jan 11 11:59 test
-rw-rw-r--+   1 satadru  satadru               5 Jan 11 11:59 tmp20
=> true
irb(main):007:0>
```

File: lib/fileutils.rb (hunk 2 above)
Comment (satmandu) [reply lib/fileutils.rb]: As per @supechicken's change I removed the entire recovery block.

File: lib/fileutils.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (satmandu) [reply lib/fileutils.rb]: Any thoughts on that @supechicken ?

[the remark this answers is not in the document]
Comment (supechicken) [reply lib/fileutils.rb]: Oh, I think you are right. This cannot handle destination without filename provided😅

File: lib/fileutils.rb [hunk 4]
```diff
@@ -1642,7 +1642,13 @@ def install(src, dest, mode: nil, owner: nil, group: nil, preserve: nil,
       st = File.stat(s)
       unless File.exist?(d) and compare_file(s, d)
         remove_file d, true
-        copy_file s, d
+        if d.to_s.end_with?('/')
```
Comment (supechicken) [inline lib/fileutils.rb:1645]: `.to_s` using with strings here is redundant

Comment (satmandu) [reply lib/fileutils.rb]: Thanks! Confirmed.

File: lib/fileutils.rb [hunk 5]
```diff
@@ -1642,7 +1642,13 @@ def install(src, dest, mode: nil, owner: nil, group: nil, preserve: nil,
...
       unless File.exist?(d) and compare_file(s, d)
         remove_file d, true
-        copy_file s, d
+        if d.end_with?('/')
+          mkdir_p d
+          copy_file s, d + File.basename(s)
+        else
+          mkdir_p File.expand_path('..', d)
+          copy_file s, d
+        end
```
Comment (uberhacker) [inline lib/fileutils.rb:1651]: I like this.  No dependency modules is great.  What do you think project maintainers?

**Comment by satmandu** [comment]
Is this acceptable @hsbt ?

**Comment by hsbt** [comment]
Thanks for your work. At least, We need to the additional tests. Can you add them?

**Comment by satmandu** [comment]
> Thanks for your work. At least, We need to the additional tests. Can you add them?

What are the additional needed tests?

**Comment by satmandu** [comment]
@hsbt What additional tests are needed?

**Comment by hsbt** [comment]
for your changes.

**Comment by satmandu** [comment]
Ah! I will modify https://github.com/ruby/fileutils/blob/master/test/fileutils/test_fileutils.rb ...

**Comment by satmandu** [comment]
Does this work?

**Comment by hsbt** [comment]
I'll fix this CI fails. It's not related with your tests.

**Comment by satmandu** [comment]
@hsbt Any chance of this getting merged?

**Pull request merged**